tb_plugin/torch_tb_profiler/profiler/memory_parser.py

```diff
@@ -0,0 +1,317 @@
+                self.processed_records_normal.append(record)
+            else:
+                self.staled_records_normal.append(record)
+        else:
+            self._update_memory_event(record, child_found)
+
+    def update_node_loop(self):
+        tree_height = 0
+        for tid, records in self.records_by_tid.items():
+            if not records:
+                return
+
+            # the traverse stack which key is the instance of node, value is the child index last visted.
+            traverse_dict = {}
+
+            child_index = 0
+            record_index = 0
+            current_node = self.tid2tree.get(tid)
+
+            if BENCHMARK_MEMORY:
+                self.processed_node[current_node] += 1
+
+            while record_index < len(records):
+                '''In the loop, one pass will process one record. The basic logic is:
+                It will search from the node that last visited since both the records and tree is ordered already
+                1. it current node contains the records, then find the exactly child which just embrace it.
+                2. otherwise, find the parent node and set the child_index, so that the parent node could continue from previous visited node.
+                3. if there is not any node contains the records, then all remaining records will be ignored.
+                '''
+                record = records[record_index]
+
+                if BENCHMARK_MEMORY and len(traverse_dict) > tree_height:
+                    tree_height = len(traverse_dict)
+
+                if current_node is None:
+                    # 3. Ignore all remaining records.
+                    logger.debug("could not find the node for tid %d, timestamp: %d, record index: %d, total records: %d" % (record.tid, record.ts, record_index, len(records)))
+                    self.staled_records.append(records[record_index])
+                    record_index += 1
+                    continue
+
+                if record.ts < current_node.start_time:
+                    # this should only happens for root node.
+                    logger.debug("record timestamp %d is less that the start time of %s" % (record.ts, current_node.name))
+                    # This record has no chance to be appended to following tree node.
+                    self.staled_records.append(record)
+                    record_index += 1
+                    continue
+                elif record.ts >= current_node.end_time:
+                    # 2. pop parent node and update the child_index accordingly.
+                    current_node = current_node.parent
+                    # next child
+                    if current_node is not None:
+                        # pop out the index of parent children. It is current node's index and plus 1 
+                        # So we can continue next node.
+                        child_index = traverse_dict.pop(current_node) + 1
+                    continue
```

Comment:
the continue will make sure the loop start from the top while. The len(current_node.children) will not hit the None exception since it only be executed when ts >= start_time and ts < end_time(that is the hidden else of 263